## Android and iOS App Analysis
This project is about looking over information about apps from the Android and iOS platforms and seeing what kinds of free, English apps are popular in the Google Play Store (Android) and the App Store (Apple).


## Taking a Look At The Dataset
The datasets that we are using for this project are 10,000 Android apps that were downloaded from the Google Play from August 2018 (link to the dataset here) and 7,000 iOS apps downloaded from App Store from July 2017 (link to the dataset here). Let's take a look at each of the datasets.


In [9]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]
    for row in dataset_slice:
        print(row)
        print('\n')
    if rows_and_columns:
        print('Number of Rows: ', len(dataset))
        print('Number of Columns: ', len(dataset[0]))

In [19]:
#Google(Android) Dataset 
from csv import reader 
open_google_file = open('C:/Users/Adity/Onedrive/Documents/Python_data_science/Datasets/googleplaystore.csv',encoding="utf8")
read_google_file = reader(open_google_file)
google_data = list(read_google_file)
google_header = google_data[0]
google_data = google_data[1:]

print(google_header)
print('\n')
explore_data(google_data, 0, 3, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of Rows:  10841
Number of Columns:  13


Based on the original Android dataset, there are a total of 10,841 rows and 13 columns. We see that there are different columns headers in the first row (i.e: App, Rating, Reviews, Type, Price, etc.) For the purposes for our analysis, the main columns that we want to be using are App, Category, Rating, Reviews, and Genres.


In [17]:
#Apple(iOS) Dataset
open_apple_file = open('C:/Users/Adity/Onedrive/Documents/Python_data_science/Datasets/AppleStore.csv',encoding="utf8")
read_apple_file = reader(open_apple_file)
apple_data = list(read_apple_file)
apple_header = apple_data[0]
apple_data = apple_data[1:]

print(apple_header)
print('\n')
explore_data(apple_data, 0, 4, True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


Number of Rows:  7197
Number of Columns:  16


Now looking at the iOS dataset, there are a total of 7,197 rows and 16 columns. There are columns headers as well with this dataset (i.e: id, track_name, price, size_bytes, etc.) For this dataset, we will be using the track_name, size_bytes, rating_count_tot, and prime_genre.


## Cleaning Up The Data
After looking at both of the datasets, we now have to do some data cleaning in order to perform the proper analysis that we want. To start, in the Android dataset, there is particular row that contains a missing value. Let's take a look the this data point.

In [20]:
print(google_data[10472])

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


In [21]:
del(google_data[10472])

Now that we have taken care of the missing values, let's go through more cleaning. We see in the Android dataset, there are duplicate entries for the same app. Let's take a look at the Instagram app and see how entries are there.

In [22]:
for app in google_data:
    name = app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


There are 4 entries of Instagram with a slight difference in the number of ratings. Let's look more closely at exactly how many duplicate apps there within the Android dataset.

In [23]:
duplicate = []
unique = []

for app in google_data: 
    name = app[0]
    if name in unique:
        duplicate.append(name)
    else:
        unique.append(name)

print('Number of duplicate apps: ', len(duplicate))
print('\n')
print('Example of duplicate apps: ', duplicate[:15])


Number of duplicate apps:  1181


Example of duplicate apps:  ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


There a total of 1,181 of duplicated apps. Now that we have a basic idea of knowing that there are duplicate apps. We need to figure out how delete these duplicates in an appropriate manner. For the purposes of our analysis, we have decided that we will be keeping the row with highest amount of rating.

In [24]:
reviews_max = {}

for app in google_data:
    name = app[0]
    n_reviews = float(app[3])
    if name in reviews_max and reviews_max[name] < n_reviews:
         reviews_max[name] = n_reviews
    elif name not in reviews_max:
         reviews_max[name] = n_reviews

print('Exact number of apps: ', len(reviews_max))

Exact number of apps:  9659


We have found the exact number of apps without any duplicates, which is a total of 9,659 apps. Now we can update the dataset

In [26]:
android_clean = []
already_added =[]

for app in google_data:
    name = app[0]
    n_reviews = float(app[3])
    if (reviews_max[name] == n_reviews) and (name not in already_added):
        android_clean.append(app)
        already_added.append(name)

explore_data(android_clean, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of Rows:  9659
Number of Columns:  13


## Removing the Non-English Apps

The next part of our data cleaning process is take out the non-English apps. So we want to only focus on English characters(A-Z), symbols (!,@,#, etc.) and numbers(0-9). These characters are encoded by the ASCII standard (each character corresponds to a number between 0-127). Once have a basic idea of how to check for English characters, we can now implement a function that help us with that using the buit-in function ord(). 

In [27]:
def is_english(string):
    for character in string:
        if ord(character) >= 127:
            return False
    return True

#Test Cases for the function '
print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

True
False
False
False


After running some case for our function, we can see that first two cases, Instagram and 爱奇艺PPS -《欢乐颂2》电视剧热播, were correct. However with the last two cases, they returned incorrectly. That is because we are looking for only one non-English character. So we need to check if there multiple non-English characters. 

In [28]:
def is_english(string):
    non_ascii = 0
    
    for character in string:
        if ord(character) >= 127:
            non_ascii += 1
        
    if non_ascii > 3:
        return False
    else:
        return True

#Test Cases for the function '
print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

True
False
True
True


Now let's update both datasets with English names. We simply use the function that we created above and add that a our new empty lists.

In [41]:
android_english = []
ios_english = []

for app in android_clean:
    if(is_english(app[0])):
        android_english.append(app)
    
for app in apple_data:
    if(is_english(app[1])):
        ios_english.append(app)

explore_data(android_english, 0, 3, True)
print('\n')
explore_data(ios_english, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of Rows:  9614
Number of Columns:  13


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+'

Now we an updated our Android dataset, which has now 9,614 rows, and the iOS dataset, which now has 6,183 rows

## Isolating the Free Apps

Now we need update our list again with free apps. The process will be similar as filtering out the non-English apps. 

In [42]:
android_final = []
ios_final = []

for app in android_english:
    price = app[7]
    if price == '0':
        android_final.append(app)
        

for app in ios_english:
    price = app[4]
    if price == '0.0':
        ios_final.append(app)

print(len(ios_final))
print(len(android_final))

3222
8864


## Finding The Most Common Genre 

Now that we finally cleaned out the data, we can now take a look and understand at what kinds of apps are downloaded in the Google Play Store and the App Store. 

In [48]:
def freq_table(dataset, index):
    table = {}
    total = 0
    for row in dataset:
        total += 1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
    
    table_percent = {}
    for key in table:
        percent = (table[key] / total ) * 100
        table_percent[key] = percent 
    
    return table_percent

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

Here we have two functions: one to create a frequency table and one to display. The first function, freq_table(), simply keeps track of the number apps for each genre and convert those counts into a percentage. Second function, display_table(), as the name suggests, shows the table 

In [49]:
#Showing Genres from the iOS dataset
display_table(ios_final, -5)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


Based on the percentages (from the prime_genre column of the App Store data set), we that the genre with most downloaded apps is Games with ~58.16%, with Entertainment with ~7.88%. Holistically, we can that the entertainment apps (games, photo and video, social networking, sports, music) have a higher percentage than the practical apps(education, shopping, utilities, productivity, lifestyle) by a huge margin. Even the top three are all from entertainment side. Just from looking at the table, we can't really make any proper observations apart from Game having higher percentage of downloaded compared to the other apps. 

Now let's take a look at the percentages for genre and category columns from the Google Play Store:

In [50]:
display_table(android_final, 1) # Category

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

In the category column, we see that the Family category has the most downloaded with Game category as the runner-up. We can see there is not a huge disparity between first and second compared to the App Store data and percentages for each are more evenly spread out than the App Store, with amount of categories in the Google Play store. We saw that there more entertainment apps downladed in the App Store than the practical apps. However, it is the converse with Google Play Store, which can be said by looking at the genres column (with tools being highest downloaded genre). As with App Store table, there is not much to go off by just looking at the tables. 

In [51]:
display_table(android_final, -4)

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

## Most Common App Downloaded in the App Store


In [52]:
genres_ios = freq_table(ios_final, -5)

for genre in genres_ios:
    total = 0
    len_genre = 0
    for app in ios_final:
        genre_app = app[-5]
        if genre_app == genre:            
            n_ratings = float(app[5])
            total += n_ratings
            len_genre += 1
    avg_n_ratings = total / len_genre
    print(genre, ':', avg_n_ratings)

Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22788.6696905016
Music : 57326.530303030304
Reference : 74942.11111111111
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 18684.456790123455
Travel : 28243.8
Shopping : 26919.690476190477
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 39758.5
Finance : 31467.944444444445
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


## Most Common App Downloaded in the Google Play Store


In [53]:
display_table(android_final, 5) # the Installs columns

1,000,000+ : 15.726534296028879
100,000+ : 11.552346570397113
10,000,000+ : 10.548285198555957
10,000+ : 10.198555956678701
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343
